### PR TITLE
📝Add line info for yaml-header-not-object and schema-not-found

### PR DESCRIPTION
**Pull request by live1206**
Add line info for `yaml-header-not-object` and `schema-not-found`

### CODE REVIEW

File: src/docfx/Errors.cs [hunk 1]
```diff
@@ -159,9 +159,10 @@ public static Error GitCloneFailed(string url, IEnumerable<string> branches)
 
         /// <summary>
         /// Yaml header defined in article.md isn't an object.
+        /// The line should always be 2 since the file should always start with "---"
         /// </summary>
-        public static Error YamlHeaderNotObject(bool isArray)
-            => new Error(ErrorLevel.Warning, "yaml-header-not-object", $"Expect yaml header to be an object, but got {(isArray ? "an array" : "a scalar")}");
+        public static Error YamlHeaderNotObject(bool isArray, string file)
+            => new Error(ErrorLevel.Warning, "yaml-header-not-object", $"Expect yaml header to be an object, but got {(isArray ? "an array" : "a scalar")}", new SourceInfo(file, 2, 1));
```
Comment (OsmondJiang) [inline src/docfx/Errors.cs]: haha

Comment (live1206) [reply src/docfx/Errors.cs]: YAML header will always be the second line :)
https://github.com/lunet-io/markdig/blob/master/src/Markdig.Tests/Specs/YamlSpecs.md

File: src/docfx/lib/json/JsonUtility.cs [hunk 2]
```diff
@@ -62,62 +62,77 @@ static JsonUtility()
...
+                return (line, column, Path.GetFileNameWithoutExtension(schema));
 
-            return null;
+            return (line, column, null);
         }
 
         /// <summary>
         /// Fast pass to read the value of $schema specified in JSON.
         /// $schema must be the first attribute in the root object.
         /// Assume input is a valid JSON. Bad input will be process though Json.NET
```
Comment (mingwli) [inline src/docfx/lib/json/JsonUtility.cs]: typo

File: src/docfx/lib/json/JsonUtility.cs [hunk 3]
```diff
@@ -62,62 +62,77 @@ static JsonUtility()
...
                 {
                     var ch = reader.Peek();
-                    if (ch == ' ' || ch == '\r' || ch == '\n' || ch == '\t')
+                    if (ch == ' ' || ch == '\t')
+                    {
+                        column += 1;
+                        reader.Read();
+                        continue;
+                    }
+                    else if (ch == '\r' || ch == '\n')
```
Comment (yufeih) [inline src/docfx/lib/json/JsonUtility.cs]: `\r` does not increment line, only `\n` increment line

File: src/docfx/lib/json/JsonUtility.cs [hunk 4]
```diff
@@ -62,62 +62,77 @@ static JsonUtility()
...
 
         /// <summary>
         /// Fast pass to read the value of $schema specified in JSON.
         /// $schema must be the first attribute in the root object.
-        /// Assume input is a valid JSON. Bad input will be process though Json.NET
+        /// Assume input is a valid JSON. Bad input will be processed through Json.NET
         /// </summary>
-        public static string ReadSchema(TextReader reader)
+        public static (int line, int column, string schema) ReadSchema(TextReader reader)
         {
```
Comment (yufeih) [inline src/docfx/lib/json/JsonUtility.cs]: If I had a second change to write this method, I would use `JsonTextReader`, that way it is simple and easy to get $schema and also line info.

File: src/docfx/lib/json/JsonUtility.cs [hunk 5]
```diff
@@ -62,65 +62,39 @@ static JsonUtility()
         /// <summary>
         /// Fast pass to read MIME from $schema attribute.
         /// </summary>
-        public static string ReadMime(TextReader reader)
+        public static (IJsonLineInfo lineInfo, string mime) ReadMime(TextReader reader)
```
Comment (OsmondJiang) [inline src/docfx/lib/json/JsonUtility.cs]: What about return SourceInfo<string> here?

File: src/docfx/lib/json/JsonUtility.cs [hunk 6]
```diff
@@ -62,65 +62,39 @@ static JsonUtility()
...
         }
 
         /// <summary>
         /// Fast pass to read the value of $schema specified in JSON.
         /// $schema must be the first attribute in the root object.
-        /// Assume input is a valid JSON. Bad input will be process though Json.NET
+        /// Assume input is a valid JSON. Bad input will be processed through Json.NET
         /// </summary>
-        public static string ReadSchema(TextReader reader)
+        public static (IJsonLineInfo lineInfo, string schema) ReadSchema(TextReader reader)
```
Comment (OsmondJiang) [inline src/docfx/lib/json/JsonUtility.cs]: Same here

File: src/docfx/lib/json/JsonUtility.cs [hunk 7]
```diff
@@ -62,65 +62,39 @@ static JsonUtility()
...
-                return Path.GetFileNameWithoutExtension(schema);
+                return (lineInfo, Path.GetFileNameWithoutExtension(schema));
 
-            return null;
+            return (lineInfo, null);
         }
 
         /// <summary>
         /// Fast pass to read the value of $schema specified in JSON.
         /// $schema must be the first attribute in the root object.
```
Comment (mingwli) [inline src/docfx/lib/json/JsonUtility.cs]: not necessary first attribute?

Comment (live1206) [reply src/docfx/lib/json/JsonUtility.cs]: outdated :)

**Pull request merged**